Feature/aml managed vnet (#43)
Azure-Samples/virtual-network-integration-recipes: src/az-aml-managed-vnet/README.md

# Azure Machine Learning Workspace in a managed virtual network (Allow Internet Outbound)

<!-- Replace "Recipe Template" title with name of the recipe. -->

## Scenario

<!-- Describe the usage scenario for this template.  Describe the challenges this recipes aims to address. -->

> This scenario aims to address the challenge of correctly configuring an Azure machine learning workspace within a Microsoft managed VNet including ensuring appropriate connectivity with common services such as Azure Storage Account, Azure Key Vault, Azure Container Registry.

Reviewer: ```suggestion
This scenario aims to address the challenge of correctly configuring an Azure Machine Learning (AML) workspace within a managed virtual network (managed VNet). This includes ensuring appropriate connectivity with common Azure services such as Azure Storage Account, Azure Key Vault, and Azure Container Registry.
```

Author: Made changes

⚠️ WARNING: This code is experimental at this stage and provided solely for evaluation purposes. It is NOT intended for production use and may contain bugs, incomplete features, or other issues. Use at your own risk, as it may undergo significant changes without notice, and no guarantees or support are provided. By using this code, you acknowledge and agree to these conditions. Consult the documentation or contact the maintainer if you have questions or concerns.

### Problem Summary

<!--Briefly describe the problme that this recipe intends to resolve or make easier. -->

> Azure machine learning workspace is composed of a number of different components: Machine Learning Studio, workspace storage account, key vault, machine learning Data Pipelines, container registry and other external data sources like Azure SQL Server, ADLS Gen2. Despite being under a single machine learning umbrella service, each of these sub-components require a slightly different VNet configuration treatment to properly isolate network traffic. For example, generally you need at least four Private Endpoints configured for a single workspace each with connecting to a different sub-component. Another example, while managed workspace are generally a single tenant service with compute resources spun up within a designated Managed VNet, data scientist vm's, azure dev ops pipelines could be multi-tenanted and therefore require provisioning a Private Endpoint within the bridge vnet in order to connect to the workspace successfully. 

Reviewer: ```suggestion
Azure machine learning workspace is composed of a number of different components: Machine Learning Studio, workspace storage account, key vault, machine learning Data Pipelines, container registry and other external data sources like Azure SQL Server, Azure Data Lake Storage Gen2 (ADLS Gen2). Despite being under a single machine learning umbrella service, each of these sub-components require a slightly different VNet configuration treatment to properly isolate network traffic. For instance, configuring at least four private endpoints is necessary for a single workspace, each connecting to a distinct sub-component. Another example pertains to managed workspaces, which are typically single-tenant services with compute resources deployed within a designated managed VNet. However, data scientist VMs and Azure DevOps pipelines could be multi-tenanted, and thus require the provisioning of a private endpoint within the bridge VNet in order to connect to the workspace successfully.
```

> In addition to this, customers will also need to ensure that traffic between the Azure machine learning workspace studio can still privately flow between the workspace components and additional Azure services such as storage external to the managed VNet. This is done through the use of Private Endpoints. Another important that one has to keep in mind is the secure integration of Azure Machine Learning with Azure DevOps pipelines and Github actions that are enabled through the bridge virtual network (VNet to access resources in the architecure diagram).

Reviewer: ```suggestion
In addition to this, customers will also need to ensure that traffic within the Azure machine learning workspace studio can still privately flow between the workspace components and additional Azure services such as external storage accounts. This is done through the use of private endpoints. Another crucial aspect to consider is the secure integration of Azure Machine Learning with Azure DevOps pipelines and GitHub actions, which is facilitated through a bridge virtual network ("VNet to access AML workspace and resources" in the architecture diagram).
```

> This recipe aims to provide developers a starting point with an IaC example of an Azure machine learning managed-vnet workspace with all sub-components correctly configured to ensure traffic stays private, while still being able to connect to common additional services such as Azure Storage Account, ADLS Gen2 and Azure Key Vault.

Reviewer: ```suggestion
This recipe aims to provide developers with a starting point, offering an Infrastructure as Code (IaC) example of an Azure Machine Learning managed VNet workspace with the required secured networking configured as described above.
```

### Architecture

<!-- Include a high-level architecture diagram of the components used in this recipe. -->

> ![architecture](./media/AML_ManagedVNet_Secure_Architecture.png)

Reviewer: Suggest removing the Azure services if they are not deployed as part of the recipe.

Also, suggest using a different font "helvetica" or "tahoma" for the text.

Reviewer: Suggest providing a brief description (few paragraphs) of what's happening in this architecture diagram.

### Recommendations

The following sections provide recommendations on when this recipe should, and should not, be used.

#### Recommended

<!-- Provide details on when usage of this recipe is recommended. -->
This recipe is recommended if the following conditions are true:

> - You want to isolate traffic at the network layer for your machine learning workspace and its sub-components.

Reviewer: ```suggestion
- You want to isolate traffic at the network layer for your AML workspace and its sub-components.
```

> - You can't have public IPs for your Azure machine learning workspace and its sub-components.

Reviewer: ```suggestion
- You can't have public IPs for your AML workspace and its sub-components.
```

#### Not Recommended

<!-- Provide details on when usage of this recipe is NOT recommended. -->
This recipe is **not** recommended if the following conditions are true:

> - Azure Machine Learning workspace is required to be accessible from the public Internet

Reviewer: ```suggestion
- The AML workspace needs to be accessible from the public internet.
```

> - Customer wants to use their own BYO virtual network for the ML workspace 

Reviewer: ```suggestion
- Customer wants to use their own virtual network for the AML workspace.
```

## Getting Started

<!-- Provide instructions on how a user would use this recipe (e.g., how they would deploy the resources). -->

### Pre-requisites

<!-- List the pre-reqs for use of this recipe (SDKs, roles/permissions, etc.) -->

> The following pre-requisites should be in place in order to successfully use this recipe:

Reviewer: ```suggestion
The following pre-requisites should be in place in order to successfully deploy this recipe:
```

- [Azure CLI](https://docs.microsoft.com/cli/azure/install-azure-cli)
- [.NET Core 7.0](https://docs.microsoft.com/dotnet/core/install/)

> - [Terraform](https://www.terraform.io/downloads.html) (Only if using Terraform)

Reviewer: ```suggestion
- [Terraform](https://www.terraform.io/downloads.html)
```

- [Terraform on Azure](https://learn.microsoft.com/en-us/azure/developer/terraform/overview)
- [Configure Terraform in Windows on Bash](https://learn.microsoft.com/en-us/azure/developer/terraform/get-started-windows-bash?tabs=bash)

> - [Azure PowerShell](https://docs.microsoft.com/powershell/azure/install-az-ps) (Only if using Azure PowerShell to deploy via Azure Bicep)

Reviewer: ```suggestion
```

- There are many options on Windows to run bash commands, including Git Bash and Windows Terminal. This article has been tested using Git Bash. Download   and install [Git Bash](https://git-scm.com/download/win).
- [Install the Azure CLI](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli-windows)
- [Install Terraform for Windows](https://learn.microsoft.com/en-us/azure/developer/terraform/get-started-windows-powershell?tabs=bash)
  - [Download Terraform](https://www.terraform.io/downloads.html)
    
    ![pwd](./media/curl_terraform_windows.png)
    
  - From the download, extract the executable to a directory of your choosing (for example, c:\terraform).
    ![pwd](./media/unzip_terraform.png)
    
  - Update your [system's global path to the executable](https://stackoverflow.com/questions/1618280/where-can-i-set-path-to-make-exe-on-windows).
    variable name - 'terraform'
  - Open a Git bash terminal window. Verify the global path configuration with the terraform command.
    ### Bash
    ```
     terraform version
    ```
     ![tf version](./media/terraform_outofdate.png)

>      - If the terraform version is out of date then use the links and steps from the "Install Terraform for Windows" to update the terraform to the latest version.

Reviewer: ```suggestion
     - If the Terraform version is outdated, follow the links and steps outlined in the [Install Terraform for Windows](https://learn.microsoft.com/en-us/azure/developer/terraform/get-started-windows-powershell?tabs=bash) guide to update Terraform to the latest version.
```

>      - For other configurations for the Terraform, refer this [link](https://learn.microsoft.com/en-us/azure/developer/terraform/quickstart-configure).

Reviewer: ```suggestion
     - For other configurations for the Terraform, refer the [quickstart](https://learn.microsoft.com/en-us/azure/developer/terraform/quickstart-configure) guide for installing and configuring Terraform.
     - Login to Azure and set the Azure subscription.
```

    ```
    pwd
    az account show
     - Login using the below command if the previous command does not return account details
    az login --use-device-code

>     az account set --subscription SUBSCRIPTION_ID

Reviewer: ```suggestion
    az account set --subscription <SUBSCRIPTION_ID>
```

    - Check user name, user type and subscription details using the command below
    az account show
     ```
    ![pwd](./media/gitbash_pwd.png)
    

>     ![pwd](./media/gitbash_azaccountshow.png)

Reviewer: ```suggestion
```

>     - Create Resource Group and make a note of it

Reviewer: ```suggestion
    - Create an Azure resource group and note down the name.
```

    ```
    az group create --location <LOCATION> --name <RESOURCE_GROUP_NAME>
    ```

    ```
    mkdir src
    ```

    ```
    cd src
    mkdir aml_managed_vnet
    cd aml_managed_vnet
    cp *.tf .
    ```
Note: For this step, the terraform scripts supplied by us has to be copied to the bash terminal at the user defined location. Subsequently, those scripts are required to be copied to the /src/aml_managed_vnet location in the same bash terminal.

    ```
     pwd
     cd /src/aml_managed_vnet
     terraform init
     terraform plan -out mvnetworkspace.tfplan
     terraform apply mvnetworkspace.tfplan

>     ```

Reviewer: These instructions are quite confusing. We should clearly explain why this copying is required?

Can we not just clone the repo? Refer https://github.com/Azure-Samples/virtual-network-integration-recipes/blob/feature/aml-managed-vnet/src/az-purview/README.md#using-terraform

Reviewer: @yogitasrivastava, also suggest mentioning that customers can adjust the default variables in the "variables.tf" file.

> ### Setting up VNet to access the Managed VNet Workspace and its resources 

Reviewer: ```suggestion
### Setting up VNet to access the managed VNet workspace and its resources

```

> In order to connect to an Azure Machine Learning (AML) workspace secured via a Managed VNet, you should create a private endpoint for the AML workspace inside a VNet where you would like to have access to the workspace (for example, inside a hub VNet). Simply follow the steps for [adding a private endpoint to a workspace](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-configure-private-link?view=azureml-api-2&tabs=cli#add-a-private-endpoint-to-a-workspace) in Azure CLI or Portal.

Reviewer: Suggest adding a section with a table to Azure resources deployed as part of the IaC deployment. Also, mention the default names of those resources.

> 

Reviewer: All the recipes in the repo actually creates the customer VNet and establishes the required private endpoint connectivity. I strongly recommend including it as part of the deployment.

Once the private endpoint is created, there are several ways that you can connect to the secured workspace. The steps in the following article use a jump box for [connecting to the workspace](https://learn.microsoft.com/en-us/azure/machine-learning/tutorial-create-secure-workspace-vnet?view=azureml-api-2#connect-to-the-workspace), which is a virtual machine in the same VNet as where you have created the private endpoint. Alternatively, you can use Azure VPN gateway for connecing on-premises networks to the VNet where the private endpoint is created, or use ExpressRoute to connect on-premises networks into the cloud over a private connection.

> In addition to adding a private endpoint for the AML workspace, a private endpoint should be created for the AML dependant resources such as Azure Blob Storage and Key Vault inside your chosen VNet to enable connectivity to these resources.

Reviewer: Again, these should be configured as part of the IaC deployment to make the sample easy to test.

> ### Testing Solution

Reviewer: In general, please following the markdown linting rules. For example, leaving spaces before and after the headers. "markdownlint" VS Code extension can help with this.

> To test Azure Machine Learning (AML) end-to-end, you can follow these steps using the **AzureML in a day notebook** from the **Samples** folder within the Notebooks section of Azure Machine Learning:

Reviewer: Where is the notebook? If it's just one notebook, I suggest copying it in the repo and add reference to it. That way, user doesn't have to go to another repo.

Having said that, please feel free to add link to the "AML rep" for reference.

Author: @oloomi Can you pls have a look?

- Create a Compute Instance:

>   - On the left navigation, select Compute and then Compute Instance. Create a new compute instance and supply a name. Keep all the defaults, expect under networking select No Public IP. Select Review & Create and wait for the deployment to be completed.

Reviewer: Can we add a GIF image to show this?

Author: @oloomi can you please have a look?

- Access the Notebook:
  - Sign in to the Azure Machine Learning studio.
  - Navigate to Notebooks and select the Samples tab.
  - Look for the AzureML in a day notebook.

>   - Open the notebook and clikc on 'Clone' to create a copy in your workspace file share.

Reviewer: I am getting the following error:

![image](https://github.com/Azure-Samples/virtual-network-integration-recipes/assets/11286537/38ade782-ddc8-4a1d-ae84-5522bbf4acb7)

I have already created "blob" PE for the storage account and AKV in the customer VNet.

Author: @oloomi can you pls have a look?
